### PR TITLE
Enable building snaps from source

### PR DESCRIPTION
**Pull request by dholbach**
Build a snap from eksctl source. @mikeroyal and I worked on this (https://github.com/weaveworks/eksctl/pull/871 and https://github.com/mikeroyal/eksctl/pull/1 are previous attempts at getting this in).

### Description

To build the snap, run

```sh
# Prep
sudo snap install snapcraft --classic
sudo snap install multipass --beta --classic

# Actual build
snapcraft
```

The resulting snap works nicely for me.

### CODE REVIEW

**Comment by dholbach** [comment]
This would need to be hooked up with CI and all, but this should be first steps for allowing us to 

- push directly to the snap store (without gatekeepers)
- get packages out to all these distros: https://docs.snapcraft.io/installing-snapd
- keep users automatically up to date
- have different [release channels](https://docs.snapcraft.io/channels)

**Comment by errordeveloper** [comment]
We should integrate this as part of our release process, goreleaser supports snaps. I believe this simply implies we need make `snap/snapcraft.yaml` part of `.goreleaser.yml`,  probably in a manner similar to how `.goreleaser.brew.yml` is done.

**Comment by dholbach** [comment]
Thanks Ilya for the feedback. I'm not sure what the best way is to hook this up with goreleaser. :man_shrugging:

**Comment by errordeveloper** [comment]
My comment above was indeed to explain the direction, did you replied
before I added extra details to it?

**Comment by dholbach** [comment]
Ah ok. Shall we do two separate PRs for this?

**Comment by errordeveloper** [comment]
@dholbach are you interested in finishing this?

**Comment by dholbach** [comment]
I was thinking we could land this piece of work first, then move on to `goreleaser` et al in a second PR. It's what I'm unfamiliar with.

**Comment by dholbach** [comment]
Can we maybe just merge this and then do second PR with whatever's required to enable the build and publication of the snap? Goreleaser has support for `snapcraft`, but it's [not complete](https://github.com/goreleaser/goreleaser/issues/914).

**Comment by dholbach** [comment]
Either we use build.snapcraft.io for this (the GH account to set this up will [need to be admin](https://github.com/canonical-web-and-design/build.snapcraft.io/issues/1191) unfortunately - to be able to set up the webhook).

If that's not an option, we could try to roll our own and do a:

```console
# Prep
sudo apt install snapd
sudo snap install snapcraft --classic
sudo snap install multipass --beta --classic

# Actual build
snapcraft
snapcraft push eksctl_<version>_amd64.snap
````

For the last step we'd need to login to the store, etc.

**Comment by dholbach** [comment]
@martina-if Thanks for assigning.

On this one I think we're blocked because there was the idea to use `goreleaser` which I don't know anything about.

What I'd suggest:

1. I update this PR
1. we merge it
1. we create a shared account on https://snapcraft.io/
1. I talk to the Snap Store maintainers, so we can take over https://snapcraft.io/eksctl from the inactive developer
1. we figure out a way to do this in CI later (using build.snapcraft.io for this unfortunately requires pretty wide access to the Github repo to set up a webhook for us and there's no manual way to do it: https://github.com/canonical-web-and-design/build.snapcraft.io/issues/1191)

The last point we haven't figured out for `fluxctl` either yet. I push snaps to the store manually every now and then. The workflow generally is:

```
# Prep
sudo snap install snapcraft --classic
sudo snap install multipass --classic

# Actual build
snapcraft

# Publishing
snapcraft push eksctl_<version>_amd64.snap --release [stable|beta|edge}
```

Let me know if it's unclear and if you have different thoughts.

**Comment by dholbach** [comment]
Updated. The PR is good to go AFAICS. Let's decouple the snap build bits from the publication in the store.

**Comment by martina-if** [comment]
Hi @dholbach thanks for updating this PR! If you think merging this is the best first step let's do that. Before that, can you please add those instructions here? https://github.com/weaveworks/eksctl/blob/master/CONTRIBUTING.md#release-process

**Comment by dholbach** [comment]
@martina-if I documented things TTBOMK. Please review.

Our options for publishing (once we took over maintenance of the snap):

- publish manually whenever we please
- publish through CI through a mechanism we can figure out ourselves
- publish using build.snapcraft.io (which would require us to at least temporarily give admin privs of the GH repo to the build service, so it can set up a webhook - not sure if that's acceptable for us)

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -228,6 +228,60 @@ This allows the message to be easier to read on GitHub as well as in various git
...
+sudo snap install multipass --classic
+```
+
+Building the snap can be done by running this command in the top-level directory:
+
+```sh
+snapcraft
+```
+
+*Note:* By default the `grade` of the snap is automatically set to `devel` and thus cannot be released to the e.g. `stable` channel of the snap. If you want to release a stable version, you want to check out the tag first, so e.g. `git checkout 0.12.` and then run `snapcraft`.
```
Comment (martina-if) [inline CONTRIBUTING.md]: nit: `git checkout 0.12.` -> `git checkout 0.12.0`

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -228,6 +228,60 @@ This allows the message to be easier to read on GitHub as well as in various git
...
+
+Where `<login-file>` was generated via `snapcraft export-login`.
+
+Then publish via:
+
+```sh
+snapcraft push eksctl_<version>_amd64.snap --release [stable,beta,candidate,edge]
+```
+
+We might want to automate some of these steps obviously. This section will likely be shorter at some stage. :-)
```
Comment (martina-if) [inline CONTRIBUTING.md]: I would remove this line or say something like:

```
TODO: automate these steps on circleci
```

**Pull request merged**